boutvecma/decoder.py

```diff
@@ -43,22 +53,94 @@ def sim_complete(self, run_info=None):
+        # the __init__ method
+        init_vars = inspect.signature(self.__init__).parameters.keys()
+        # Assuming that the __init__ does `self.var = var`, we can
+        # just return a dict of everything
+        try:
+            return {var: getattr(self, var) for var in init_vars}
+        except AttributeError as e:
+            # Let's try to give a helpful error message if this does go wrong
+            match = re.search(r"no attribute '(.*)'", str(e))
+            if match is None:
+                # Oops, something else
+                raise e
+            varname = match.group(1)
+            raise AttributeError(
+                f"'{varname}' doesn't seem to be a member of this instance.\nDid you miss 'self.{varname} = {varname}' in '__init__'?"
+            )
+
+    @staticmethod
+    def element_version():
+        return "0.1.0"
+
+
+class SimpleBOUTDecoder(BaseBOUTDecoder):
+    """Just collects individual variables at the last timestep"""
+
+    def __init__(self, target_filename=None, variables=None):
+        """
+        Parameters
+        ==========
+        variables: iterable or None
+            Iterable of variables to collect from the output. If None, return everything
+        """
+        super().__init__(target_filename=target_filename)
+
+        # TODO: check it's an iterable or otherwise sensible type?
+        self.variables = variables
+
+    def parse_sim_output(self, run_info=None, *args, **kwargs):
+        df = self.get_outputs(run_info)
 
-        # For now, just return the variable itself
-        # TODO: add analysis step
         return {
-            variable: df[variable][-1, ...].values.flatten().tolist()
+            variable: flatten_dataframe_for_JSON(df[variable][-1, ...])
             for variable in self.variables
         }
 
-    def get_restart_dict(self):
+    @staticmethod
+    def element_version():
+        return "0.1.0"
+
+
+class SampleLocationBOUTDecoder(BaseBOUTDecoder):
+    """Samples variables at single points or areas"""
+
+    def __init__(self, sample_locations):
+        """
+        sample_locations should be a list of dicts, each like:
+
+            {
+              "variable": "T",  # variable to sample
+              "output_name": "T_centre",  # name in the output
+              # Following are the indices to sample
+              "x": 0,
+              "y": 50,
+              "z":0
+            }
+        """
+        super().__init__()
+
+        self.sample_locations = sample_locations
+
+    def parse_sim_output(self, run_info=None, *args, **kwargs):
+        df = self.get_outputs(run_info)
+
+        samples = copy.deepcopy(self.sample_locations)
+
         return {
-            "target_dir": self.target_dir,
-            "target_filename": self.target_filename,
-            "variables": self.variables,
+            variable.pop("output_name"): flatten_dataframe_for_JSON(
+                df[variable.pop("variable")][variable]
```

Comment:
This exact code is probably a little bit too gross and fragile, but something like this is probably what we want as a default/bog-standard decoder. xarray dataframes can be index with a dict of coordinates like here, so this is very flexible: the whole field, the last time point and all space, all time and one spatial point, are all easily supported with the same decoder. At the minute it just supports indexing in index-space, so I'll weigh in with John about how to support indexing in coordinate-space in order to support e.g. resolution scans.